libparsec/crates/client_connection/src/authenticated_cmds/sse.rs

```diff
@@ -0,0 +1,242 @@
+            .and_then(|raw| {
+                T::api_load_response(raw.as_ref()).map_err(|_| ConnectionError::BadContent)
+            })
+            .map(SSEResponseOrMissedEvents::Response),
+
+        _ => return std::task::Poll::Pending,
+    };
+
+    std::task::Poll::Ready(message.map(|message| SSEEvent {
+        id: event.id,
+        retry: event.retry,
+        message,
+    }))
+}
+
+pub(crate) fn handle_sse_error(error: EventStreamError<reqwest::Error>) -> ConnectionError {
+    match error {
+        EventStreamError::Utf8(_) | EventStreamError::Parser(_) => ConnectionError::BadContent,
+        EventStreamError::Transport(e) => ConnectionError::NoResponse(Some(e)),
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SSEConnectionError {
+    #[error(transparent)]
+    Transport(reqwest::Error),
+    #[error("Invalid content type {0:?}")]
+    InvalidContentType(reqwest::header::HeaderValue),
+    #[error("Invalid status code {0}")]
+    InvalidStatusCode(reqwest::StatusCode),
+}
+
+impl From<SSEConnectionError> for ConnectionError {
+    fn from(value: SSEConnectionError) -> Self {
+        match value {
+            SSEConnectionError::Transport(err) => ConnectionError::NoResponse(Some(err)),
+            SSEConnectionError::InvalidContentType(_) => ConnectionError::BadContent,
+            // All statuses except 200
+            SSEConnectionError::InvalidStatusCode(code) => match code.as_u16() {
+                415 => ConnectionError::BadContent,
+                // TODO: cannot  access the response headers here...
+                // 422 => crate::error::unsupported_api_version_from_headers(
+                //     resp.headers(),
+                // )),
+                460 => ConnectionError::ExpiredOrganization,
+                461 => ConnectionError::RevokedUser,
+                // We typically use HTTP 503 in the tests to simulate server offline,
+                // so it should behave just like if we were not able to connect
+                503 => ConnectionError::NoResponse(None),
+                _ => ConnectionError::InvalidResponseStatus(code),
+            },
+        }
+    }
+}
+
+pub struct RateLimiter {
+    /// How much time we tried to backoff between a reset
+    attempt: usize,
+    desired_duration: Option<Duration>,
+}
+
+impl RateLimiter {
+    pub const fn new() -> Self {
+        Self {
+            attempt: 0,
+            desired_duration: None,
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.attempt = 0;
+    }
+
+    pub fn set_desired_duration(&mut self, duration: Duration) {
+        self.desired_duration = Some(duration);
+    }
+}
+
+impl RateLimiter {
+    pub async fn wait(&mut self) {
+        let duration_to_wait = self.get_duration_to_wait();
+        self.attempt += 1;
+        libparsec_platform_async::sleep(duration_to_wait).await;
+    }
+
+    fn get_duration_to_wait(&self) -> Duration {
+        self.desired_duration.unwrap_or(match self.attempt {
```

Comment:
Two issues here:
1 - we want to use `max(desired_duration, duration_from_attempt)` (so that if `desired_duration = 5s` we will nevertheless end up waiting 60s before each attempt)
2 - if attempt is zero we always return 0s (given it means we don't have a previous failed attempt, so we should try to connect without delay)

Reply:
And we should rename `desired_duration` to something like `min_retry_wait`

Reply:
2. I've considered the following scenario:
   
   The server will restart and indicate the client that it is restarting and that it need to way N seconds then close the stream.
   The client should wait N seconds and not retry directly.

Reply:
:+1: